### PR TITLE
Refactor lexer functions

### PR DESCRIPTION
**Pull request by yt2b**
I refactored following lexer functions.
* `lex_number`
* ~~`take_number`~~
* `at_exponent`
* `is_identifier_continuation`

What I have done are as follows.
* using match expression instead of if expression
* reducing the nesting of match expression

### CODE REVIEW

**Comment by youknowone** [comment]
@bonjune could you review this PR?

**Comment by bonjune** [comment]
> @bonjune could you please review this PR?

I'll look into it today.

File: compiler/parser/src/lexer.rs [hunk 1]
```diff
@@ -439,10 +438,9 @@ where
     fn take_number(&mut self, radix: u32) -> Option<char> {
         let take_char = Lexer::<T>::is_digit_of_radix(self.window[0], radix);
 
-        if take_char {
-            Some(self.next_char().unwrap())
-        } else {
-            None
+        match take_char {
+            true => Some(self.next_char().unwrap()),
+            _ => None,
```
Comment (bonjune) [inline compiler/parser/src/lexer.rs:443]: I like the idea of pattern matching, but if expression should be preferred to match expression when we are dealing with a boolean value.

**Review by bonjune — CHANGES_REQUESTED** [review]
I like how the pattern matching here improves readability.
But, as I left a comment, if expression should be preferred for boolean values.

File: compiler/parser/src/lexer.rs (hunk 1 above)
Comment (yt2b) [reply compiler/parser/src/lexer.rs:443]: I'll restore if expression in `take_number`

**Comment by yt2b** [comment]
@bonjune 
Code fix completed.

**Review by bonjune — APPROVED** [review]
Good!
Have a good day. Thank you.

File: compiler/parser/src/lexer.rs (hunk 1 above)
Comment (DimitrisJim) [reply compiler/parser/src/lexer.rs]: note that this could probably use [`bool.then`](https://doc.rust-lang.org/std/primitive.bool.html#method.then)

Comment (yt2b) [reply compiler/parser/src/lexer.rs]: Thank you for the adivse.

**Comment by yt2b** [comment]
@bonjune 
Could you please review again?

**Review by bonjune — APPROVED** [review]
Looks good.

P.S.

https://doc.rust-lang.org/std/primitive.bool.html#method.then_some

I read bool's two method implementation: `bool::then` and `bool::then_some`, whose behaviors are subtly different. The key thing is here: the former lazily evaluated the passed closure while the latter one evaluates passed value eagerly. In this case, `bool::then` is a proper choice because we don't to evaluate the function only when `take_char` is true. (And you made the right decision, I think)

cc @youknowone

**Review by youknowone — APPROVED** [review]
Thank you!

**Pull request merged**